File: Python/isomorphic_strings.py
# https://leetcode.com/problems/isomorphic-strings/description/
from utils.examination import Examiner, Testcase, Logger


class Solution:
    def isIsomorphic(self, s: str, t: str) -> bool:
        if len(s) != len(t):
            return False
        s_to_t: dict[str, str] = {}
        t_to_s: dict[str, str] = {}
        for i in range(len(s)):
            if s[i] not in s_to_t and t[i] not in t_to_s:
                s_to_t[s[i]] = t[i]
                t_to_s[t[i]] = s[i]
            if s[i] not in s_to_t or t[i] not in t_to_s:
                return False
            if s[i] != t_to_s[t[i]]:
                return False
        return True

def examine()->None:
    Examiner(Solution(), 'isIsomorphic').run((
        Testcase(('paper', 'title'), True),
        Testcase(('egg', 'add'), True),
        Testcase(('egg', 'adda'), False),
        Testcase(('egg', 'addb'), False),
        Testcase(('egg', 'ad'), False),
        Testcase(('eg', 'add'), False),
        Testcase(('ege', 'add'), False),
        Testcase(('eggp', 'add'), False),
    ))

File: TypeScript/solutions/isomorphic-strings.ts
// https://leetcode.com/problems/isomorphic-strings/description/
import { Examiner, Testcase } from '../utils';


function isIsomorphic(s: string, t: string): boolean {
    if (s.length !== t.length) return false;
    const sToT = new Map<string, string>();
    const tToS = new Map<string, string>();
    for (let i = 0; i < s.length; i++) {
        const sc = s[i]!, tc = t[i]!;
        if (!sToT.has(sc) && !tToS.has(tc)) { sToT.set(sc, tc); tToS.set(tc, sc); }
        if (!sToT.has(sc) || !tToS.has(tc)) return false;
        if (sc !== tToS.get(tc)!) return false;
    }
    return true;
}

export function examine(): void {
    new Examiner(isIsomorphic).run([
        new Testcase(['paper', 'title'], true),
        new Testcase(['egg', 'add'], true),
        new Testcase(['egg', 'adda'], false),
        new Testcase(['egg', 'addb'], false),
        new Testcase(['egg', 'ad'], false),
        new Testcase(['eg', 'add'], false),
        new Testcase(['ege', 'add'], false),
        new Testcase(['eggp', 'add'], false),
    ]);
}
